### ISSUE TITLE
fix(gguf): harden cursor bounds and tensor-size arithmetic against malformed files

### ISSUE THREAD

**Issue by geisten** [geisten/geistlib#332]
Severity: HIGH

## What happened

The GGUF reader performs several bounds checks only after pointer arithmetic or cursor advancement:

- `src/io/gguf_reader.c:135,142,152,165`: checks based on `c->p + len`
- `src/io/gguf_reader.c:184-198,222`: `skip_value` advances first and validates later
- `src/io/gguf_reader.c:117-123`: tensor dimension multiplication is unchecked
- `src/io/gguf_reader.c:261-265,411-414`: `general.alignment == 0` can reach modulo arithmetic
- `src/io/gguf_reader.c:431-456`: tensor byte size and `data_offset + offset + nbytes` can overflow

For an attacker-controlled length, `&c->p[len]` is **not** safer than `c->p + len`: C defines the subscript expression in terms of the same pointer addition. Forming a pointer beyond one-past the mapped object is already undefined behavior, even when it is immediately compared with `end`.

## Expected

Use subtraction-based checks before any pointer update, ideally through one cursor helper:

```c
if (len > (size_t)(c->end - c->p))
    return GEIST_ERR_FORMAT;
const uint8_t *src = c->p;
c->p += len;
```

Also use checked add/multiply helpers for dimensions, tensor byte counts, offsets, and alignment rounding. Reject zero/non-power-of-two alignment according to the accepted GGUF contract.

## Acceptance criteria

- No parser check forms an untrusted out-of-range pointer before validation.
- All dimension, byte-count, offset, and alignment arithmetic is checked.
- Malformed/truncated GGUF fixtures cover huge lengths, overflowing dimensions/offsets, and zero alignment.
- ASan/UBSan and the malformed-I/O suite pass.
- Add fuzz seeds for each rejected arithmetic class.

## Mandatory C23 implementation rules

- In every modified array API, place all length/capacity parameters before the arrays they describe; use VLA contracts such as `T data[static len]` where the non-null/minimum-size precondition is real.
- Use C23 `bool`, `true`, `false`, and `nullptr`/`nullptr_t` instead of integer booleans or `NULL`. Replace constant macros with typed `constexpr` objects when compile-time and ABI requirements permit.
- Use `[[nodiscard]]` for returned status/size values that must be checked. Add `restrict` only when non-aliasing is proven for every caller. Use `typeof`/`auto` only when they improve clarity without hiding ownership or changing code generation.
- Perform checked `size_t` arithmetic before allocation, indexing, alignment, or pointer advancement. For cursors, validate against `(size_t)(end - p)` before advancing; do not use `p + len` or `&p[len]` as the bounds check.
- Preserve public ABI unless the ticket explicitly includes a versioned compatibility migration.

## Mandatory performance and quality gates

- Benchmark GGUF open/load time and the standard inference sweep before and after every parser change that touches the valid-file path.
- Valid GGUF files must expose identical tensor metadata/data and produce identical logits within existing tolerances.
- Run malformed-input tests, ASan, UBSan, and fuzz seeds after every parser arithmetic/cursor batch.
- No repeatable load-time, inference-throughput, RSS, or quality regression is acceptable; keep expensive checks outside kernel hot loops.

**Comment by geisten** [comment]
Fixed in 44a4868 (PR #344), released in **v0.10.2**.

**What changed.** Every bounds test now goes through one cursor helper that subtracts two pointers into the same mapping and takes bytes only after the length fits — `p + len` for an attacker-chosen `len` is undefined before the comparison runs, and the compiler may drop the check. `skip_value` no longer advances first and validates afterwards.

The arithmetic around it goes through the new `src/base/checked.h` (C23 `ckd_*` where `<stdckdint.h>` exists, the long-standing builtins where it does not): tensor dimensions, byte counts, `data_offset + offset + nbytes`, and the alignment round-up. Zero and non-power-of-two `general.alignment` are **rejected** rather than replaced by the default — a file that disagrees with us about padding puts its tensor data somewhere we would not look.

**Acceptance criteria.**
- [x] No parser check forms an untrusted out-of-range pointer before validation
- [x] All dimension, byte-count, offset and alignment arithmetic checked
- [x] Malformed fixtures for huge lengths, overflowing dimensions/offsets, zero alignment
- [x] ASan/UBSan and the malformed-I/O suite pass
- [x] Fuzz seeds per rejected arithmetic class — eight added to `test_io_malformed_unit`: length one byte under `UINT64_MAX`, array count `UINT64_MAX`, zero alignment, non-power-of-two alignment, overflowing dimension product, overflowing byte count, wrapping offset

**Performance gate.** Load time on the 3.1 GB reference model is unchanged, 0.02 s best-of-7 before and after — parsing is O(tensors) and mmap dominates. Full inference sweep in PR #344.
